src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java

```diff
@@ -0,0 +1,671 @@
+            for (Pair<Integer, RejectBefore> pair : rangeMaps)
+            {
+                int storeId = pair.left;
+                RejectBefore rejectBefore = pair.right;
+
+                if (rejectBefore == null)
+                    continue;
+
+                rejectBefore.foldlWithBounds(
+                    (txnId, ds, start, end) -> ds.row(storeId, decompose(start), decompose(end)).column("txn_id", decompose(txnId)),
+                    dataSet,
+                    ignore -> false
+                );
+            }
+            return dataSet;
+        }
+    }
+
+    public static class TxnBlockedByTable extends AbstractVirtualTable
+    {
+        enum Reason { Self, Txn, Key }
+
+        protected TxnBlockedByTable(String keyspace)
+        {
+            super(parse(keyspace,
+                        "Accord Transactions Blocked By Table" ,
+                        "CREATE TABLE txn_blocked_by (\n" +
+                           format("txn_id %s,\n", TXN_ID_TYPE_STRING) +
+                        "  command_store_id int,\n" +
+                        "  depth int,\n" +
+                           format("blocked_by %s,\n", TXN_ID_TYPE_STRING) +
+                        "  reason text,\n" +
+                        "  save_status text,\n" +
+                           format("execute_at %s,\n", TIMESTAMP_TYPE_STRING) +
+                           format("key %s,\n", ROUTING_KEY_TYPE_STRING) +
+                        "  PRIMARY KEY (txn_id, command_store_id, depth, blocked_by, reason)" +
+                        ')'));
+        }
+
+        @Override
+        public DataSet data(DecoratedKey partitionKey)
+        {
+            TxnId id = compose(partitionKey.getKey());
+            List<CommandStoreTxnBlockedGraph> shards = AccordService.instance().debugTxnBlockedGraph(id);
+
+            SimpleDataSet ds = new SimpleDataSet(metadata());
+            for (CommandStoreTxnBlockedGraph shard : shards)
+            {
+                Set<TxnId> processed = new HashSet<>();
+                process(ds, shard, processed, id, 0, id, Reason.Self, null);
+                // everything was processed right?
+                if (!shard.txns.isEmpty() && !shard.txns.keySet().containsAll(processed))
+                    throw new IllegalStateException("Skipped txns: " + Sets.difference(shard.txns.keySet(), processed));
+            }
+
+            return ds;
+        }
+
+        private void process(SimpleDataSet ds, CommandStoreTxnBlockedGraph shard, Set<TxnId> processed, TxnId userTxn, int depth, TxnId txnId, Reason reason, Runnable onDone)
+        {
+            if (!processed.add(txnId))
+                throw new IllegalStateException("Double processed " + txnId);
+            CommandStoreTxnBlockedGraph.TxnState txn = shard.txns.get(txnId);
+            if (txn == null)
+            {
+                Invariants.checkState(reason == Reason.Self, "Txn %s unknown for reason %s", txnId, reason);
```

Comment:
this is actually a bug... I was going to fix in CASSANDRA-20054 but was asked to drop to speed up the review.  We have a concept of "historic transactions", so those come back as `null` here.

The fix for it is

```
-                Invariants.checkState(reason == Reason.Self, "Txn %s unknown for reason %s", txnId, reason);
+                Invariants.checkState(reason != Reason.Self, "Self Txn %s is unknown", txnId);
+                // Self txn depends on this txn, but it is not known...
+                ds.row(userTxn.toString(), shard.storeId, depth, txn.txnId.toString(), reason.name());
+                ds.column("save_status", "<unknown>");
```

Reply:
Just `txn_d.toString()` I assume? Making this change breaks all the tests in `AccordDebugKeyspaceTest` though.